Introduce System2 cache.

david-german-tri opened the pull request:
Systems can use the cache to store and retrieve computed values.  The cache will automatically invalidate entries whose dependencies have changed.

Sending this out for review with slightly incomplete test coverage, since I anticipate some design back-and-forth.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3135)
<!-- Reviewable:end -->

sherm1 commented:
Sorry for the delayed review. I will try to get to this in depth Monday but it is going to be hectic since I have a new employee starting (Paul Mitiguy). I do have a couple of quick design comments:
- cache values should not be deleted when invalidated for several reasons: (a) the previous value is very handy for updating complicated cache entries like the updated value for a delay buffer (no need to recreate it from scratch), (b) for "mode" variables it is very useful to swap the mode value with its cached update value at near-zero cost, (c) the out-of-date value is essential in some expert uses (like inside integrators and multibody internals -- I'm not talking about approximation here; it can be the right value to use), and (d) heap reallocation is expensive and shouldn't be done in inner loops (which is where the cache action is taking place).
- a map is too slow for cache entry lookup. The main design principle is to make cache access as close to ordinary variable access as possible. There is time for an index operation, check validity flag, return reference to value and little else. It all has to be inline. I have profiled this extensively in Simbody and Drake is performing identical computations so I expect that to transfer.

It is very important to me that caching work efficiently. I will be very disappointed if programmers start writing workarounds for speed! I know physical simulation programmers very well and I know they'll hack around anything they think is too slow (I'm looking at you, @amcastro-tri!). I also know they won't get the caching code right and we'll start getting subtly wrong answers, which would be a disaster.

I would be delighted to allocate some dynamics group time to working on the cache -- we will shortly have some new muscle in the form of very competent dynamicists. 

---

Review status: 0 of 18 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-:-KPBiq-hyuYWRFRVFplV:b8xvj5n)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
All of that is good feedback - thanks!  I think the resulting changes are not so large.  I'll make updates later today; feel free to hold off on further review until then.

david-german-tri commented:
Done, PTAL!

liangfok commented:
Reviewed 3 of 17 files at r1.
Review status: 3 of 17 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/basic_vector.h, line 30 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJfWM8Tln_EhU5RsT_:-KPJfWM97APaZoBhxuzs:b36d28e) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/basic_vector.h#L30)):*
> ```objc
>                       typename Eigen::NumTraits<T>::Real>::quiet_NaN())) {}
> 
>   /// Constructs a BasicVector with the specified &p data.
> ```

Should that be `@p`? The following page did not mention `&` as a possible symbol for denoting parameters:

https://www.stack.nl/~dimitri/doxygen/manual/commands.html

---

*[drake/systems/framework/context.h, line 79 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJgEe_RTmBxnUbjDoc:-KPJgEeaCDxL3-VxzJY1:b5t3ytu) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/context.h#L79)):*
> ```objc
>   /// To prevent cache corruption, callers MUST NOT write on the returned
>   /// pointer after writing to the cache or passing this Context to a System
>   /// that might write on the cache.
> ```

Is there a way to write to the cache other than through this returned pointer?

---

*[drake/systems/framework/context.h, line 173 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJgR1rc68VyBWHvj8f:-KPJgR1rc68VyBWHvj8g:b-t0qoed) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/context.h#L173)):*
> ```objc
>   /// Allocates @p n new input ports, not connected to anything. This function
>   /// may only be called once, because modifying the number of input ports
>   /// after context creation would destroy the semantics of cache tickets.
> ```

BTW I believe we've decided to not use doxygen blocks for private code.

---

*[drake/systems/framework/context_base.h, line 69 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJgdxKENGa1G2ozB5L:-KPJgdxKENGa1G2ozB5M:b-xpp9kq) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/context_base.h#L69)):*
> ```objc
> 
>   /// Invalidates all cache lines that depend on the time.
>   /// User code should not call this function.
> ```

BTW since these methods are not intended to be publicly visible, I wonder if this would be a good use case of the `friend` keyword, which would allow these methods to be private except to the friend class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I'm still seeing the deleted value and map -- is there something that didn't get pushed?

---

Reviewed 3 of 17 files at r1.
Review status: 3 of 17 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/cache.cc, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPKaouXudfx0RpMP4th:-KPKaouXudfx0RpMP4ti:b7myvbi) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/cache.cc#L22)):*
> ```cpp
>   // added after the fact, the dependency map is guaranteed acyclic.
>   for (const CacheTicket& dependency : dependencies) {
>     cache_dependencies_[dependency].insert(ticket);
> ```

BTW, I found this confusing due to dependency/dependent being an awful pair of near-opposite words in English. I suggest using "prerequisite" exclusively to mean "thing I depend on" -- it significantly lightens the cognitive load. The "pre" is a lovely indicator of who must get evaluated first!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-:-KPKcBQc7xjUePrh3nuK:b-n3bopo)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Huh.  Reviewable is showing me `std::vector<CacheEntry> store_;`.  Not so for you?

---

Review status: 3 of 17 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-:-KPNh4vJ1nu2YIybtnt0:b-i5zdcy)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Oh, I see what you mean now.  Actually done this time, and added a test, PTAL!

---

Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/framework/basic_vector.h, line 30 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJfWM8Tln_EhU5RsT_:-KPPRL8T_w72p4RglXR4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/basic_vector.h#L30)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Should that be `@p`? The following page did not mention `&` as a possible symbol for denoting parameters:
> 
> https://www.stack.nl/~dimitri/doxygen/manual/commands.html
</details>
Done.

---

*[drake/systems/framework/cache.cc, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPKaouXudfx0RpMP4th:-KPPRjAwEjNcv9S-RrKQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/cache.cc#L22)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, I found this confusing due to dependency/dependent being an awful pair of near-opposite words in English. I suggest using "prerequisite" exclusively to mean "thing I depend on" -- it significantly lightens the cognitive load. The "pre" is a lovely indicator of who must get evaluated first!
</details>
Done.

---

*[drake/systems/framework/context.h, line 79 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJgEe_RTmBxnUbjDoc:-KPPSlMtSQ2UXGGhfjvr:b-gyylzw) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/context.h#L79)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Is there a way to write to the cache other than through this returned pointer?
</details>
I've tried to improve the comment, PTAL.

---

*[drake/systems/framework/context.h, line 173 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJgR1rc68VyBWHvj8f:-KPPSrac2llCHcjh9WkH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/context.h#L173)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW I believe we've decided to not use doxygen blocks for private code.
</details>
Done.

---

*[drake/systems/framework/context_base.h, line 69 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJgdxKENGa1G2ozB5L:-KPPT--Vbzl2tQvhmY0I:b-xtt2hi) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/context_base.h#L69)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW since these methods are not intended to be publicly visible, I wonder if this would be a good use case of the `friend` keyword, which would allow these methods to be private except to the friend class.
</details>
Eh, actually, I've deleted the comment.  I can't think of a good reason for users to call these functions, but they won't blow anything up.

`friend` is an interesting suggestion, but the challenge with `friend`s, of course, is that they overshare.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-:-KPPTqBZqfeFyf3RkHp6:b8fvo1y)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I am working my way through this -- may need a f2f + whiteboard! Will formulate some specific questions.

---

Reviewed 4 of 17 files at r1.
Review status: 4 of 17 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/cache.h, line 18 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPu7t8v7gQq9AOuHMy:-KPPu7t8v7gQq9AOuHMz:bgf7nhp) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L18)):*
> ```objc
> typedef int CacheTicket;
> 
> class CacheEntry {
> ```

BTW, needs a class comment.

---

*[drake/systems/framework/cache.h, line 37 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPuDiOS5zwOwbFibkz:-KPPuDiOS5zwOwbFibl-:bbk06a2) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L37)):*
> ```objc
>  private:
>   bool is_valid_{false};
>   std::unique_ptr<AbstractValue> value_;
> ```

Should add list of dependents here, or better consider having this implement a "value source" abstraction that deals with downstream dependents and can be used for other value sources such as inputs, time, state, etc.

---

*[drake/systems/framework/cache.h, line 55 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPjd3L8_aqh0YEYt4u:-KPPjd3L8_aqh0YEYt4v:b-9dm1k3) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L55)):*
> ```objc
>   /// Creates a new cache ticket, which will be invalidated whenever any of
>   /// the dependencies are invalidated.
>   CacheTicket MakeCacheTicket(std::set<CacheTicket> prerequisites);
> ```

How are non-cache entry prerequisites dealt with (like states, parameters, inputs)? Are you planning to assign cache tickets to them?

---

*[drake/systems/framework/cache.h, line 76 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPh30GzFAgcPoG6Ldh:-KPPgpnemlHdiaey4lG4:b-tuhfq4) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L76)):*
> ```objc
>   // For each cache ticket, the set of other cache tickets that are invalidated
>   // when it is invalidated. This graph is directed and acyclic.
>   std::map<CacheTicket, std::set<CacheTicket>> cache_dependencies_;
> ```

BTW, `cache_dependents_` would be better. "dependencies" is ambiguous here and can be synonymous with prerequisites.

---

*[drake/systems/framework/cache.h, line 76 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPiaetoRFattdUVAtZ:-KPPiaetoRFattdUVAt_:b4hym2l) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L76)):*
> ```objc
>   // For each cache ticket, the set of other cache tickets that are invalidated
>   // when it is invalidated. This graph is directed and acyclic.
>   std::map<CacheTicket, std::set<CacheTicket>> cache_dependencies_;
> ```

`cache_dependents_` is indexed identically to `store_` so the `dependents` set can be added to CacheEntry and kept in the store. That will make it easier to see everything that's involved in a CacheEntry object, and get rid of the map.

---

*[drake/systems/framework/context_base.h, line 73 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPsrLAlPh4N99fIEbZ:-KPPsrLAlPh4N99fIEb_:b-f42153) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/context_base.h#L73)):*
> ```objc
>   /// Invalidates all cache lines that depend on the state.
>   /// TODO(david-german-tri): Provide finer-grained invalidation.
>   virtual void InvalidateState() = 0;
> ```

BTW, yes we'll need at least q,v,z, xd but ideally allow subgroupings of those. Also individual mode variables.

---

*[drake/systems/framework/diagram_context.h, line 189 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPvqOhko-noibVXBrj:-KPPvqOhko-noibVXBrk:b-ci2ked) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/diagram_context.h#L189)):*
> ```objc
> 
>   /// Propagates the invalidation of time to all subsystems.
>   /// User code should not call this function.
> ```

I think you told Liang you removed this restriction. I agree that it is fine to expose invalidation in the API -- it is particularly useful for for testing.

---

*[drake/systems/framework/diagram_context.h, line 190 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPqQlwsxwr1CmR1BRA:-KPPqQlwsxwr1CmR1BRB:bqeklqp) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/diagram_context.h#L190)):*
> ```objc
>   /// Propagates the invalidation of time to all subsystems.
>   /// User code should not call this function.
>   void InvalidateTime() override {
> ```

BTW, many (most?) subsystems don't depend on time (so far: adder, gain, constant source, even integrator). It would be better for those that do to say so rather than assume they all do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-:-KPQDIbQMbL2CpWjddEV:bx8nmto)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 6 of 17 files at r1, 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/context.h, line 79 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPJgEe_RTmBxnUbjDoc:-KPQMUbsu17fdxKoia37:b-g6llej) ([raw file](https://github.com/robotlocomotion/drake/blob/ebc5c0c462539b8ebb9285b9b84c344011957b73/drake/systems/framework/context.h#L79)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I've tried to improve the comment, PTAL.
</details>
Makes sense. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 13 of 17 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/cache.h, line 18 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPu7t8v7gQq9AOuHMy:-KPSlMESOf73C8k-K_xM:b-42wx1b) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L18)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, needs a class comment.
</details>
Done.  Also added `namespace internal`.  

I wanted it to be a struct, but MSVC errors instead of deleting the copy constructor from structs with non-copyable elements.  Then I wanted to forward-declare it and put the definition in the .cc file, but MSVC requires the entire definition for classes that are used as template arguments.  Now that it's a class, at least we're getting some good use of it from your other suggestions.

---

*[drake/systems/framework/cache.h, line 37 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPuDiOS5zwOwbFibkz:-KPSnaNEOmAgfZqD3vTJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L37)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Should add list of dependents here, or better consider having this implement a "value source" abstraction that deals with downstream dependents and can be used for other value sources such as inputs, time, state, etc.
</details>
Done.

---

*[drake/systems/framework/cache.h, line 55 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPjd3L8_aqh0YEYt4u:-KPSmUFLlMxLb_B50CGW:bhcur4n) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L55)):*
> Are you planning to assign cache tickets to them?

Yes.

---

*[drake/systems/framework/cache.h, line 76 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPh30GzFAgcPoG6Ldh:-KPSn05TTHNr0gj-fsdr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L76)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, `cache_dependents_` would be better. "dependencies" is ambiguous here and can be synonymous with prerequisites.
</details>
Done.

---

*[drake/systems/framework/cache.h, line 76 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPiaetoRFattdUVAtZ:-KPSnazNf8vNFsQVveGv:b7mqpyl) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/cache.h#L76)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> `cache_dependents_` is indexed identically to `store_` so the `dependents` set can be added to CacheEntry and kept in the store. That will make it easier to see everything that's involved in a CacheEntry object, and get rid of the map.
</details>
Oh, nice.  Done.

---

*[drake/systems/framework/context_base.h, line 73 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPsrLAlPh4N99fIEbZ:-KPSo_gbxpc_t9J4mZyl:bpd2urc) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/context_base.h#L73)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, yes we'll need at least q,v,z, xd but ideally allow subgroupings of those. Also individual mode variables.
</details>
Made comment more specific.

---

*[drake/systems/framework/diagram_context.h, line 189 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPvqOhko-noibVXBrj:-KPSocAWXE5HxaMoWHYz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/diagram_context.h#L189)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I think you told Liang you removed this restriction. I agree that it is fine to expose invalidation in the API -- it is particularly useful for for testing.
</details>
Done.

---

*[drake/systems/framework/diagram_context.h, line 190 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPqQlwsxwr1CmR1BRA:-KPSomUJIl-ZVXCezrwH:b-e2yj3e) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/diagram_context.h#L190)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, many (most?) subsystems don't depend on time (so far: adder, gain, constant source, even integrator). It would be better for those that do to say so rather than assume they all do.
</details>
Interesting.  Why does this matter?  For systems that have no time-dependent cache lines, this should be pretty quick: each system in the Diagram gets visited for invalidation, resulting in a call to `cache_->Invalidate(time_ticket)`.  The time ticket will have no dependents, so that's the end of it.

We could prune the worst case from O(N) to O(1) in the size of the diagram with preregistration: nothing in the diagram is time-dependent, so we know just to stop here.  That seems like a lot of machinery when we don't even believe invalidation is a hotspot, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Some comments.

---

Reviewed 3 of 11 files at r2.
Review status: 16 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/framework/diagram_context.h, line 190 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPqQlwsxwr1CmR1BRA:-KPUJt4Mj_vAUVWawrc5:b-ysn16t) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/diagram_context.h#L190)):*
> Interesting. Why does this matter?

It matters because the actual dependency graph is a truly fundamental property of the computation we're performing and not to be able to represent that is, in my view, a fundamental design flaw. I feel that this is a sign that there is an important missing abstraction at the moment. You had part of that in an earlier implementation that I liked (and stole for sys3), called a `ValueListener` -- that's a beautiful abstraction for any object whose value is dependent on some prerequisite values. The other one should be something like `ValueSource` and would be implemented by any object that can provide a value and notify listeners that they are out of date. These are extremely important concepts and I think they deserve to be first class interfaces.

With those concepts, Time would be a ValueSource (along with Inputs, Parameters, state variables, and more), and any listener who cares would register with Time. I am very sad not to see these fundamental concepts directly represented in the code. I don't see why there has to be any argument about whether the performance is actually going to matter for some special case -- let's first make sure we can represent and implement the fundamental dependency concepts just once and then re-use them wherever they occur. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-:-KPUNMQ7Y3_AjRY2EWoS:b-rua8el)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/cache.h, line 71 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUOBC4TtC7Yn9fFQ5L:-KPUOBC532JPBFGmPc0v:b-tnuqml) ([raw file](https://github.com/robotlocomotion/drake/blob/2744447fdf923544b1997dba8ed43951d9f19fb6/drake/systems/framework/cache.h#L71)):*
> ```objc
> 
>   /// Takes ownership of a cached item, and returns a bare pointer to the item.
>   AbstractValue* Set(CacheTicket ticket, std::unique_ptr<AbstractValue> value);
> ```

BTW under normal operation updating a cache entry will be done by assignment to the underlying concrete object, not by replacement of its abstract value. That avoids inner-loop heap activity, prevents surprise changes to the underlying concrete type, and permits cheap updating of complex cache entries such as delay tables.

In certain cases there is a need for a value *swap* though (typically for mode variables that are paired with an identical cache entry for updating every time step). I'm not sure what is the best API for that -- could be just a `release()` method combined with this take-ownership Set method but might be nicer with an explicit Swap(). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 13 of 17 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/cache.h, line 71 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUOBC4TtC7Yn9fFQ5L:-KPUV4l9M10_aCo0PV2O:b-n8rifw) ([raw file](https://github.com/robotlocomotion/drake/blob/2744447fdf923544b1997dba8ed43951d9f19fb6/drake/systems/framework/cache.h#L71)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW under normal operation updating a cache entry will be done by assignment to the underlying concrete object, not by replacement of its abstract value. That avoids inner-loop heap activity, prevents surprise changes to the underlying concrete type, and permits cheap updating of complex cache entries such as delay tables.
> 
> In certain cases there is a need for a value *swap* though (typically for mode variables that are paired with an identical cache entry for updating every time step). I'm not sure what is the best API for that -- could be just a `release()` method combined with this take-ownership Set method but might be nicer with an explicit Swap(). 
</details>
Makes sense.  I'd mildly prefer `Swap` to `release`, so I've added it.

---

*[drake/systems/framework/diagram_context.h, line 190 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPqQlwsxwr1CmR1BRA:-KPUSOHJgaN3GAY2x6sa:b-szlgj6) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/diagram_context.h#L190)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> > Interesting. Why does this matter?
> 
> It matters because the actual dependency graph is a truly fundamental property of the computation we're performing and not to be able to represent that is, in my view, a fundamental design flaw. I feel that this is a sign that there is an important missing abstraction at the moment. You had part of that in an earlier implementation that I liked (and stole for sys3), called a `ValueListener` -- that's a beautiful abstraction for any object whose value is dependent on some prerequisite values. The other one should be something like `ValueSource` and would be implemented by any object that can provide a value and notify listeners that they are out of date. These are extremely important concepts and I think they deserve to be first class interfaces.
> 
> With those concepts, Time would be a ValueSource (along with Inputs, Parameters, state variables, and more), and any listener who cares would register with Time. I am very sad not to see these fundamental concepts directly represented in the code. I don't see why there has to be any argument about whether the performance is actually going to matter for some special case -- let's first make sure we can represent and implement the fundamental dependency concepts just once and then re-use them wherever they occur. 
</details>
OK, I agree that Systems need to report their sparsity/functional-form matrix.  @bradking is working on that, and once we have it, I agree we might as well use it here for a minor optimization.  So, perhaps we're on the same page.  Added a TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
This was painful for me to review. I am uncomfortable with this bottom-up approach to developing the API. I found it hard to see how this fits into the complete picture. What does the API look like for a real System to request context resources, initialize values properly, manage and update cache entries, export arbitrary computations to output ports, produce derivatives, updates, witness functions, etc. while avoiding unnecessary copying and computation? Can I express dependencies on subgroupings of state variables? Will the present code be sufficient for all those uses? Will the proposed sorting scheme work (a) for internal computations, (b) for discrete sampling, (c) when there are multiple directed graphs in a diagram, (d) at all? What is wrong with using a uniform dataflow method that is guaranteed to work? Why are their special cases for different kinds of value sources? How does a user encapsulate a complex System and return it as a self-contained object for someone else to use? I can't tell and I fear it will take a long time and a lot of rework to get through the partial list in #2738. 

Some comments/rants below.

I'm going to work on the solver now to get a top-down look; maybe that will help me grasp the full scheme.

---

Reviewed 2 of 17 files at r1, 4 of 11 files at r2, 1 of 4 files at r3, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/cache.h, line 68 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUm9Ary2Ou5s-jrGKG:-KPUm9Ary2Ou5s-jrGKH:bwel22r) ([raw file](https://github.com/robotlocomotion/drake/blob/71e1e9fc5de6c7e136be189cfeb83b38b47da88c/drake/systems/framework/cache.h#L68)):*
> ```objc
>   /// Creates a new cache ticket, which will be invalidated whenever any of
>   /// the dependencies are invalidated.
>   CacheTicket MakeCacheTicket(std::set<CacheTicket> prerequisites);
> ```

comment: dependencies->prerequisites

---

*[drake/systems/framework/cache.h, line 70 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUp7sJ7UNQWbobxZau:-KPUp7sKuJXIm_GZBI0k:b3ftn1v) ([raw file](https://github.com/robotlocomotion/drake/blob/71e1e9fc5de6c7e136be189cfeb83b38b47da88c/drake/systems/framework/cache.h#L70)):*
> ```objc
>   CacheTicket MakeCacheTicket(std::set<CacheTicket> prerequisites);
> 
>   /// Invalidates and deletes the value for @p ticket, and all lines that
> ```

BTW doesn't delete the value anymore, just invalidates.
BTW lines->entries.

---

*[drake/systems/framework/cache.h, line 72 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUmFHf3Z6oKWgWyrS8:-KPUmFHf3Z6oKWgWyrS9:b-y8jt7q) ([raw file](https://github.com/robotlocomotion/drake/blob/71e1e9fc5de6c7e136be189cfeb83b38b47da88c/drake/systems/framework/cache.h#L72)):*
> ```objc
>   /// Invalidates and deletes the value for @p ticket, and all lines that
>   /// depend on it.
>   void Invalidate(CacheTicket ticket);
> ```

I am still unclear why this shouldn't invalidate *all* dependents rather than just *cache entry* dependents, meaning some other code has to find dependent outputs (and I'm fuzzy about whether you plan also to special-case derivatives, updates, witness functions, and other pre-defined System computations). I don't get, for example, how the present scheme allows me to take some intermediate cached computations (say ArticulatedBodyInertias) and expose them in an OutputPort, while continuing to compute downstream computations like accelerations (out the Derivative method and an OutputPort) and  reaction forces (another OutputPort). Do I have to copy the gigantic ArticulatedBodyInertia cache entry into some special OutputPort-only value location?

Basically I just don't understand the urge to special case these things at all, *especially* output port values, which are just one among many computations and need exactly the same cache & dependency management support as everything else.

---

*[drake/systems/framework/cache.h, line 79 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUq2lJvx1CyMux0Jue:-KPUq2lJvx1CyMux0Juf:b-702nlf) ([raw file](https://github.com/robotlocomotion/drake/blob/71e1e9fc5de6c7e136be189cfeb83b38b47da88c/drake/systems/framework/cache.h#L79)):*
> ```objc
>   /// Returns the cached item for the given ticket, or nullptr if the item
>   /// has been invalidated.
>   AbstractValue* Get(CacheTicket ticket) const;
> ```

BTW, I'm not sure how this would be used. If the canonical formula is `check validity; calculate if necessary; return value` then it will be necessary to have a mutable pointer to the value object in order to calculate into it, but this return nullptr.

---

*[drake/systems/framework/cache.h, line 83 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUfc-lds95v_C6Qww4:-KPUfc-lds95v_C6Qww5:b-eechqr) ([raw file](https://github.com/robotlocomotion/drake/blob/71e1e9fc5de6c7e136be189cfeb83b38b47da88c/drake/systems/framework/cache.h#L83)):*
> ```objc
>   /// Returns the cached item for the given ticket, or nullptr if the item
>   /// has been invalidated. Replaces it with @p value.
>   std::unique_ptr<AbstractValue> Swap(CacheTicket ticket,
> ```

Should return the AbstractValue object regardless of whether it is valid -- this is going to be used for re-use of the physical object as tentative calculations become permanent parts of the trajectory after trial steps are accepted.

---

*[drake/systems/framework/context.h, line 20 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUrX7c50ZJe0Qt2x3k:-KPUrX7c50ZJe0Qt2x3l:bw0gx7j) ([raw file](https://github.com/robotlocomotion/drake/blob/71e1e9fc5de6c7e136be189cfeb83b38b47da88c/drake/systems/framework/context.h#L20)):*
> ```objc
> 
> template <typename T>
> struct ContextShape {
> ```

Oh dear. I really don't like this. If I understand it correctly it is a roundabout way to express that there are value sources in the context upon which various computations may depend (though excluding cache entries upon which those computations may also depend). "ContextShape" does not strike me as a meaningful abstraction. I suggest something more straightforward like a ValueSource interface from which suitable sources could inherit. Then as a value consumer (ValueListener?) I could point at Time, or an Input, or a Parameter, or a State variable subgroup, or a mode variable, or a cache entry, and say "I depend on that ValueSource".

---

*[drake/systems/framework/context_base.h, line 29 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUkEI1Xx7zRoRqRrYp:-KPUkEI2P_ixREvnyrCc:b-syhuz6) ([raw file](https://github.com/robotlocomotion/drake/blob/71e1e9fc5de6c7e136be189cfeb83b38b47da88c/drake/systems/framework/context_base.h#L29)):*
> ```objc
> ///           scalar.
> template <typename T>
> class ContextBase {
> ```

BTW, sorry for out-of-place comment but this class should be named `Context` so that we can use that name in the API. Then if necessary it can have concrete subclasses LeafContext and DiagramContext -- users shouldn't have to think about that.

---

*[drake/systems/framework/context_base.h, line 58 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPUkkrV9PAOWtr5rI6z:-KPUkkrV9PAOWtr5rI7-:b1fd9b7) ([raw file](https://github.com/robotlocomotion/drake/blob/71e1e9fc5de6c7e136be189cfeb83b38b47da88c/drake/systems/framework/context_base.h#L58)):*
> ```objc
> 
>   /// Returns writable access to the State.
>   /// Implementations should invalidate all cache lines that depend on state.
> ```

BTW, I think "cache entries" would be a better term here than "lines" -- there is no natural sequence here as there is in a hardware cache. Cache entries are independent.

---

*[drake/systems/framework/diagram_context.h, line 190 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-KPPqQlwsxwr1CmR1BRA:-KPUgsiT966zebEZQvHp:b-4e6vx8) ([raw file](https://github.com/robotlocomotion/drake/blob/a44cf3b49a0b7a6b3050499c95a102f4af437b7e/drake/systems/framework/diagram_context.h#L190)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> OK, I agree that Systems need to report their sparsity/functional-form matrix.  @bradking is working on that, and once we have it, I agree we might as well use it here for a minor optimization.  So, perhaps we're on the same page.  Added a TODO.
</details>
OK, though I don't get how FunctionalForm applies here. This is a simple wiring diagram issue IMO -- Time is like an input port and an Adder doesn't have such a port. I don't understand why it has to be complicated. But I'm happy to wait to see what you have in mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3135#-:-KPUtQShNW4rsmFqreK9:buzw0v0)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
OK, I'll just close this PR, as it sounds like we could squabble about it for a long time, and it isn't necessary for any of the short-term priorities @RussTedrake has articulated.
